src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,483 @@
+        # the states are: OK, REPLICATING, STUCK, SUSPENDED, WAITING_APPROVAL, INJECT
+        if state == 'OK':
+            # log that the rule has been satisfied and add that RSE to the currentRSEs (unique)
+            msg = f"monitoring task {uuid}, container {pname} under the rule ID {rid}"
+            msg += f" targeting RSEs {rses} in rucio account {rucioAccount} is completely available"
+            logger.info(msg)
+            for rse in rses:
+                if rse not in doc['currentRSEs']:
+                    doc['currentRSEs'].append(rse)
+                    modify = True
+                    msg = f"update currentRSEs with {rse}"
+                    report.addEntry('monitoring', uuid, msg)
+        else:
+            # calculate the rule completion based on the 3 locks_* field
+            sumOfLocks = rdoc['locks_ok_cnt'] + rdoc['locks_replicating_cnt'] + rdoc['locks_stuck_cnt']
+            completion = rdoc['locks_ok_cnt'] / sumOfLocks
+            msg = f"monitoring task {uuid}, container {pname} under the rule ID {rid}"
+            msg += f" targeting RSEs {rses} in rucio account {rucioAccount} has"
+            msg += f" a fraction completion of {completion}"
+            logger.info(msg)
+            for rse in rses:
+                if rse in doc['currentRSEs']:
+                    doc['currentRSEs'].remove(rse)
+                    modify = True
+                    msg = f"delete rse {rse} from currentRSEs"
+                    report.addEntry('monitoring', uuid, msg)
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info(f"monitoring task {uuid}, update {pname}")
+        mgr.updatePileup(doc)
+        msg = f"update pileup {pname}"
+        report.addEntry('monitoring', uuid, msg)
+    else:
+        logger.info("monitoring task {uuid}, processed without update for {pname}")
+
+
+def inactiveTask(doc, spec):
+    """
+    Perform single inactive task over provided MSPileup document
+
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    logger = spec['logger']
+    pname = doc['pileupName']
+    kwargs = {'scope': 'cms', 'account': rucioAccount}
+    if isinstance(rucioClient, Client):
+        inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+        rules = list(rucioClient.list_replication_rules(inputArgs))
+    else:
+        rules = rucioClient.listDataRules(pname, **kwargs)
+    modify = False
+
+    for rdoc in rules:
+        # make a Rucio call to delete that rule id
+        rid = rdoc['id']
+        msg = f"inactive task {uuid}, container: {pname} for Rucio account {rucioAccount}"
+        msg += f", delete replication rule {rid}"
+        logger.info(msg)
+        if isinstance(rucioClient, Client):
```

Comment:
It has not been mentioned in the requirements, but let me write it here before we forget it.

We should have a mechanism to run the "rule deletion" in dry-run mode, such that we don't really delete any rule, but just log that it would have been deleted.

Reply:
the question is do you want dry-run mode parameter for a class (and object will always be in dry-mode), or do you want dry-mode per API?

Reply:
Dry-run mode should be service-wide. However, we only need to bypass (dry-run) actual Rucio write operations, which are create and delete rules AFAICT.
Maybe in that scenario we would also want to not update the document in MongoDB, but if we update with no actual modifications that's equally fine.

Reply:
Added in this commit https://github.com/dmwm/WMCore/pull/11499/commits/5f49cd4944d904aa6956334251e4c48e6927eb85